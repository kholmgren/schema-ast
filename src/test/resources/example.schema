/** tenant/user represents a user */
definition tenant/user {}

/** tenant/group represents a group **/
definition tenant/group {
    /** member is a member of a tenant/group, which can be a tenant/user or the membership of another tenant/group */
    relation member: tenant/user | tenant/group#member
}

/** tenant/document represents a document */
definition tenant/document {
    /** writer is a writer of the document */
    relation writer: tenant/user | tenant/group#member

    /** reader is a reader of the document */
    relation reader: tenant/user | tenant/group#member

    /** write indicates which tenant/user can write to the document */
    permission write = writer

    /** read indicates which tenant/user can read the document */
    permission read = reader + write
}