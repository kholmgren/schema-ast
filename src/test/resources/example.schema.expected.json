{
  "@type" : "schema",
  "definitions" : [ {
    "@type" : "definition",
    "name" : "tenant/user"
  }, {
    "@type" : "definition",
    "name" : "tenant/group",
    "relations" : [ {
      "@type" : "relation",
      "name" : "member",
      "expression" : {
        "@type" : "or",
        "left" : {
          "@type" : "ref",
          "ref" : "tenant/user"
        },
        "right" : {
          "@type" : "groupRef",
          "ref" : "tenant/group",
          "member" : "member"
        }
      }
    } ]
  }, {
    "@type" : "definition",
    "name" : "tenant/document",
    "relations" : [ {
      "@type" : "relation",
      "name" : "writer",
      "expression" : {
        "@type" : "or",
        "left" : {
          "@type" : "ref",
          "ref" : "tenant/user"
        },
        "right" : {
          "@type" : "groupRef",
          "ref" : "tenant/group",
          "member" : "member"
        }
      }
    }, {
      "@type" : "relation",
      "name" : "reader",
      "expression" : {
        "@type" : "or",
        "left" : {
          "@type" : "ref",
          "ref" : "tenant/user"
        },
        "right" : {
          "@type" : "groupRef",
          "ref" : "tenant/group",
          "member" : "member"
        }
      }
    } ],
    "permissions" : [ {
      "@type" : "permission",
      "name" : "write",
      "expression" : {
        "@type" : "ref",
        "ref" : "writer"
      }
    }, {
      "@type" : "permission",
      "name" : "read",
      "expression" : {
        "@type" : "union",
        "left" : {
          "@type" : "ref",
          "ref" : "reader"
        },
        "right" : {
          "@type" : "ref",
          "ref" : "write"
        }
      }
    } ]
  } ]
}